rust/src/ssh/parser.rs

```diff
@@ -92,10 +128,89 @@ named!(pub ssh_parse_record<SshRecordHeader>,
         padding_len: be_u8 >>
         msg_code: be_u8 >>
         take!((pkt_len-2) as usize) >>
-        (SshRecordHeader{pkt_len, padding_len, msg_code})
+        (SshRecordHeader{pkt_len: pkt_len,
+            padding_len: padding_len,
+            msg_code: MessageCode::from_u8(msg_code)})
     )
 );
 
+#[derive(Debug,PartialEq)]
+pub struct SshPacketKeyExchange<'a> {
+    pub cookie: &'a[u8],
+    pub kex_algs: &'a [u8],
+    pub server_host_key_algs: &'a [u8],
+    pub encr_algs_client_to_server: &'a [u8],
+    pub encr_algs_server_to_client: &'a [u8],
+    pub mac_algs_client_to_server: &'a [u8],
+    pub mac_algs_server_to_client: &'a [u8],
+    pub comp_algs_client_to_server: &'a [u8],
+    pub comp_algs_server_to_client: &'a [u8],
+    pub langs_client_to_server: &'a [u8],
+    pub langs_server_to_client: &'a [u8],
+    pub first_kex_packet_follows: u8,
+}
+
+use md5::compute;
+
+const SSH_HASSH_STRING_DELIMITER_SLICE: [u8; 1] = [b';'];
+
+impl<'a> SshPacketKeyExchange<'a> {
+    pub fn generate_hassh(&self, hassh_string: &mut Vec<u8>, hassh: &mut Vec<u8>, to_server: &bool) {
+        let slices = if *to_server { 
+            [self.kex_algs, &SSH_HASSH_STRING_DELIMITER_SLICE,
+             self.encr_algs_server_to_client, &SSH_HASSH_STRING_DELIMITER_SLICE,
+             self.mac_algs_server_to_client, &SSH_HASSH_STRING_DELIMITER_SLICE,
+             self.comp_algs_server_to_client]}
+        else {
+            [self.kex_algs, &SSH_HASSH_STRING_DELIMITER_SLICE,
+             self.encr_algs_client_to_server, &SSH_HASSH_STRING_DELIMITER_SLICE,
+             self.mac_algs_client_to_server, &SSH_HASSH_STRING_DELIMITER_SLICE,
+             self.comp_algs_client_to_server]
+        };
+        // reserving memory
+        hassh_string.reserve_exact(slices.iter().fold(0, |acc, x| acc + x.len()));
+        // copying slices to hassh string
+        slices.iter().for_each(|&x| hassh_string.extend_from_slice(x)); 
+        hassh.extend(format!("{:x?}", compute(&hassh_string)).as_bytes());
+    }
+}
+
+named!(parse_string<&[u8]>, do_parse!(
```

Comment:
For reference : https://securitylab.github.com/research/network-integers-are-hot-lava

Reply:
I tested for the incorrect length of fields and incorrect strings in kex message, parsing worked fine. Added unit test.

Reply:
Your code seems fine to me indeed with respect to this :-)
But then, the first referenced patch seemed fine to me (and to the author of the post)
So, thanks for double checking